GeneratorInterface/Core/interface/ConcurrentGeneratorFilter.h

```diff
@@ -299,24 +315,39 @@ namespace edm {
   }
 
   template <class HAD, class DEC>
-  void ConcurrentGeneratorFilter<HAD, DEC>::globalBeginLuminosityBlockProduce(LuminosityBlock& lumi,
-                                                                              EventSetup const& es) const {
-    // need one of the streams to finish
+  std::shared_ptr<gen::GenLumiCache<HAD, DEC>> ConcurrentGeneratorFilter<HAD, DEC>::globalBeginLuminosityBlock(
+      edm::LuminosityBlock const&, edm::EventSetup const&) const {
+    //need one of the streams to finish
     while (useInLumi_.load() == nullptr) {
     }
+
+    if (nInitializedInGlobalLumi_ < nGlobalBeginRuns_) {
+      while (!streamEndRunComplete_.load()) {
+      }
+      nInitializedInGlobalLumi_ = nGlobalBeginRuns_;
+    }
+
+    auto lumiCache = std::make_shared<gen::GenLumiCache<HAD, DEC>>();
+    lumiCache->useInLumi_ = useInLumi_.load();
+    return lumiCache;
+  }
+
+  template <class HAD, class DEC>
+  void ConcurrentGeneratorFilter<HAD, DEC>::globalBeginLuminosityBlockProduce(LuminosityBlock& lumi,
+                                                                              EventSetup const& es) const {
     initLumi(useInLumi_, lumi, es);
     std::unique_ptr<GenLumiInfoHeader> genLumiInfoHeader(useInLumi_.load()->hadronizer_.getGenLumiInfoHeader());
     lumi.put(std::move(genLumiInfoHeader));
+    useInLumi_.store(nullptr);
   }
 
   template <class HAD, class DEC>
   void ConcurrentGeneratorFilter<HAD, DEC>::streamBeginLuminosityBlock(StreamID id,
                                                                        LuminosityBlock const& lumi,
                                                                        EventSetup const& es) const {
-    if (useInLumi_ != this->streamCache(id)) {
-      initLumi(this->streamCache(id), lumi, es);
-    } else {
-      useInLumi_.store(nullptr);
+    gen::GenStreamCache<HAD, DEC>* streamCachePtr = this->streamCache(id);
+    if (this->luminosityBlockCache(lumi.index())->useInLumi_ != streamCachePtr) {
+      initLumi(streamCachePtr, lumi, es);
```

Comment:
Just for clarity, should have have an `else` which sets the lumi cache useInLum_ to nullptr? It isn't strictly needed as the lumi cache will go away at the end of the luminosity block. But it might make the logic easier to understand and avoid possible uses of that cache later.

Reply:
On second thought, resetting the lumi cache `useInLumi_` here would require the variable to become an atomic, which really isn't worth it.

Reply:
This transition will run multiple times (once per stream), possibly concurrently. useInLumi_ in the lumi cache is not atomic. Only reading it here is OK, but writing would create a data race. I could make it atomic, but it does not seem worth adding the overhead of an atomic. I'm not sure whether your proposed extra code clarifies or just leads the reader to wonder why one set it null... As is, this lumi cache value is written to in one line of code and read in only this one line code.

Reply:
I should have refreshed my window. I didn't see your second comment before my reply.